R/parallel.R

```diff
@@ -3,47 +3,87 @@
+##' `hipercow_resources`. Each process here would be use a single core.
+##'
+##' You can also call `hipercow_parallel` with `cores_per_process`,
+##' to make hipercow launch as many processes as it can with
+##' each process having the number of cores you request, with
+##' the total cores being at most what you requested with
+##' `hipercow_resources`.
+##'
+##' For example, you could request 32 cores with `hipercow_resources`,
+##' and then calll `hipercow_parallel` with `cores_per_process = 4`,
+##' and hipercow will create a local cluster with 8 processes, each
+##' of which reporting `4` cores if that process calls
+##' `hipercow_parallel_get_cores`.
+##'
+##' If you did the same with `cores_per_process = 5`, hipercow would
+##' create 6 local processes, each reporting `5` cores, and
+##' two cores would be effectively unallocated.
+##'
+##' Here are some brief examples; see `vignette("parallel")` for more
+##' details. In each example, we are looking up the process id (to
+##' show that different processes are being launched), and asking each
+##' process how many cores it should be using.
+##'
+##' For using the `future` package:
 ##'
 ##' ```
 ##' resources <- hipercow_resources(cores = 4)
 ##' id <- task_create_expr(
-##'   furrr::future_map(1:4, ~Sys.getpid()),
+##'   furrr::future_map(1:4,
+##'     ~c(Sys.getpid(), hipercow_parallel_get_cores()),
 ##'   parallel = hipercow_parallel("future"),
 ##'   resources = resources)
+##'
 ##' ```
 ##'
 ##' where `furrr` must be provisioned using [hipercow_provision].
-##'
-##' For use with `parallel`:
+##' Here is an equivalent example with `parallel`:
 ##'
 ##' ```
+##' resources <- hipercow_resources(cores = 4)
 ##' id <- task_create_expr(
-##'   parallel::clusterApply(NULL, 1:4, function(x) Sys.getpid()),
+##'   parallel::clusterApply(NULL, 1:4, function(x)
+##'     c(Sys.getpid(), hipercow_parallel_get_cores()),
 ##'   parallel = hipercow_parallel("parallel"),
 ##'   resources = resources)
 ##' ```
 ##'
+##'
 ##' @title Specify parallel use of cores
 ##'
 ##' @param method The parallel method that hipercow will prepare.
 ##' Three options are available: the `future` package, the `parallel`
-##' page, or `NULL`, the default, will do nothing. See the details for
+##' package, or `NULL`, the default, will do nothing. See the details for
 ##' examples.
 ##'
+##' @param cores_per_process The number of cores allocated to each
+##' process when launching a local cluster using one of the parallel
+##' methods. By default, this will be `1`. See details.
+##'
 ##'
 ##' @return A list containing your parallel configuration.
 ##'
 ##' @export
-hipercow_parallel <- function(method = NULL) {
+hipercow_parallel <- function(method = NULL, cores_per_process = 1L) {
+  assert_scalar_integer(cores_per_process)
+
   if (!is.null(method) &&
       (!method %in% c("future", "parallel"))) {
     cli::cli_abort(c(
-      "Parallel method {method} unknown.",
+      'Parallel method "{method}" unknown.',
```

Comment:
```suggestion
      "Parallel method '{method}' unknown",
```

Normally we quote this way around, and no full stops

Reply:
Done